Source/WebCore/dom/ElementTextDirection.cpp

```diff
@@ -0,0 +1,314 @@
+{
+    for (RefPtr child = element.firstChild(); child; ) {
+        // Specs: Skip bdi, script, style nodes.
+        if (child->hasTagName(bdiTag) || child->hasTagName(scriptTag) || child->hasTagName(styleTag)) {
+            child = NodeTraversal::nextSkippingChildren(*child, &element);
+            continue;
+        }
+
+        if (auto* childElement = dynamicDowncast<Element>(*child)) {
+            // Specs: Skip text form controls
+            // Specs: Skip the element whose dir attribute is not in the undefined state
+            if (childElement->isTextField() || elementHasValidTextDirectionState(*childElement)) {
+                child = NodeTraversal::nextSkippingChildren(*child, &element);
+                continue;
+            }
+        }
+
+        // Specs: If child is a slot element whose root is a shadow root,
+        // then return the directionality of that shadow root's host.
+        if (auto* childSlotElement = dynamicDowncast<HTMLSlotElement>(*child)) {
+            if (RefPtr childHost = childSlotElement->shadowHost())
+                return computeTextDirection(*childHost, elementTextDirectionState(*childHost));
+        }
+
+        // Specs: If descendant is a Text node and its text direction is not null,
+        // then return it.
+        if (child->isTextNode()) {
+            if (auto direction = computeTextDirectionFromText(child->textContent(true)))
+                return direction;
+        }
+
+        child = NodeTraversal::next(*child, &element);
+    }
+
+    return std::nullopt;
+}
+
+// Specs: https://html.spec.whatwg.org/multipage/dom.html#auto-directionality
+static std::optional<TextDirection> computeTextDirectionOfSlotElement(const HTMLSlotElement& slotElement)
+{
+    // Specs: If element is a slot element whose root is a shadow root
+    // and element's assigned nodes are not empty, then return the
+    // directionality of this slot.
+    if (!slotElement.isInShadowTree())
+        return computeContainedTextAutoDirection(slotElement);
+
+    auto* nodes = slotElement.assignedNodes();
+    if (!nodes)
+        return computeContainedTextAutoDirection(slotElement);
+
+    for (auto& child : *nodes) {
+        // Specs: If child is a Text node and its text direction is
+        // not null, then return it.
+        if (child->isTextNode()) {
+            if (auto direction = computeTextDirectionFromText(child->textContent(true)))
+                return direction;
+        }
+
+        if (auto* element = dynamicDowncast<Element>(child.get())) {
+            // Specs: If child direction is not null, then return child direction.
+            if (auto direction = computeAutoTextDirection(*element))
+                return direction;
+        }
+    }
+
+    return std::nullopt;
+}
+
+// Specs: https://html.spec.whatwg.org/multipage/dom.html#auto-directionality
+std::optional<TextDirection> computeAutoTextDirection(const Element& element)
```

Comment:
We should really call this function computeAutoDirectionality to match the spec.

Reply:
Done.